### ISSUE TITLE
4097 Error

### ISSUE THREAD

**Issue by timmrogers** [google/cups-connector#68]
I've made some progress - Go and cups-connector are installed and seem to be working nicely. CUPS is installed and I've configured several printers. 

But I am getting the following error message:

```
trogers@debian:~$ connector
E0914 23:30:21.437047    8860 connector.go:33] Google Cloud Print CUPS Connector version DEV-linux
Google Cloud Print CUPS Connector version DEV-linux
E0914 23:30:24.203739    8860 connector.go:119] Ready to rock as proxy 'OHS-CloudPrint'
Ready to rock as proxy 'OHS-CloudPrint'
E0914 23:31:24.221903    8860 printermanager.go:207] Sync failed while calling GetPrinters(): Failed to call cupsDoRequest() [IPP_OP_CUPS_GET_PRINTERS]: Failed to connect to CUPS server /var/run/cups/cups.sock:631 because 4097 Error in the pull function.
```

Any suggestions?

Thanks again.

**Comment by jacobmarble** [comment]
I have never see "Error in the pull function" before. A quick search looks like this is a GnuTLS error. I don't know why TLS would be necessary when communicating with a local socket.

Maybe turn off encryption? You can do that [in a config file, or an environment variable](http://www.cups.org/documentation.php/doc-2.1/api-cups.html#cupsEncryption). Possible values are listed in `cups/http.h`:
```
  HTTP_ENCRYPTION_IF_REQUESTED,	/* Encrypt if requested (TLS upgrade) */
  HTTP_ENCRYPTION_NEVER,		/* Never encrypt */
  HTTP_ENCRYPTION_REQUIRED,		/* Encryption is required (TLS upgrade) */
  HTTP_ENCRYPTION_ALWAYS		/* Always encrypt (SSL) */
```

I often test the connector like this; you could do something similar:
```
$ HTTP_SERVER=servername connector -logtostderr
```

**Comment by timmrogers** [comment]
Thank you.

Any clues here? I see that connector seems to think that "encrypting
REQUIRED". Was that perhaps some configuration issue on my end with CUPS? I
just did a standard install via tasksel.

trogers@debian:~$ HTTP_SERVER=servername connector -logtostderr

E0915 09:38:34.955707    9795 connector.go:33] Google Cloud Print CUPS
Connector version DEV-linux

Google Cloud Print CUPS Connector version DEV-linux

I0915 09:38:37.033222    9795 core.go:86] connected to CUPS server
/var/run/cups/cups.sock:631 encrypting REQUIRED

I0915 09:38:37.033544    9795 connector.go:88] SNMP enabled

I0915 09:38:37.046853    9795 avahi.go:257] Avahi client running.

I0915 09:38:37.811751    9795 printermanager.go:219] Synchronizing
printers, stand by

I0915 09:38:37.951327    9795 printermanager.go:252] Printers are already
in sync; there are 3

I0915 09:38:37.961472    9795 printermanager.go:115] Registered Computer
locally

I0915 09:38:37.996405    9795 printermanager.go:115] Registered McGehee
locally

I0915 09:38:38.014933    9795 printermanager.go:115] Registered MacDonald
locally

E0915 09:38:38.015363    9795 connector.go:119] Ready to rock as proxy
'OHS-CloudPrint'

Ready to rock as proxy 'OHS-CloudPrint'

I0915 09:39:38.016315    9795 printermanager.go:219] Synchronizing
printers, stand by

E0915 09:39:38.035697    9795 printermanager.go:207] Sync failed while
calling GetPrinters(): Failed to call cupsDoRequest()
[IPP_OP_CUPS_GET_PRINTERS]: Failed to connect to CUPS server
/var/run/cups/cups.sock:631 because 4097 Error in the pull function.

**Comment by jacobmarble** [comment]
I meant that as an example of setting an environment variable as a prefix to the command. Try this:
```
$ CUPS_ENCRYPTION=NEVER connector -logtostderr
E0915 09:56:44.658812   34134 connector.go:33] Google Cloud Print CUPS Connector version DEV-darwin
Google Cloud Print CUPS Connector version DEV-darwin
I0915 09:56:46.545988   34134 core.go:86] connected to CUPS server /private/var/run/cupsd:631 encrypting NEVER
I0915 09:56:46.968970   34134 printermanager.go:219] Synchronizing printers, stand by
I0915 09:56:47.017874   34134 printermanager.go:252] Printers are already in sync; there are 2
E0915 09:56:47.017998   34134 connector.go:119] Ready to rock as proxy 'test-laptop'
Ready to rock as proxy 'test-laptop'
```

**Comment by timmrogers** [comment]
Following this HOWTO to create a self-signed certificate

http://edoceo.com/howto/cups

seemed to make the connector happier - but I still get the same error.

```
trogers@debian:~$ HTTP_SERVER=servername connector -logtostderr
E0915 09:51:57.917146    9910 connector.go:33] Google Cloud Print CUPS Connector version DEV-linux
Google Cloud Print CUPS Connector version DEV-linux
I0915 09:51:59.571457    9910 core.go:86] connected to CUPS server /var/run/cups/cups.sock:631 encrypting REQUIRED
I0915 09:51:59.571778    9910 connector.go:88] SNMP enabled
I0915 09:51:59.585000    9910 avahi.go:257] Avahi client running.
I0915 09:52:00.476055    9910 printermanager.go:219] Synchronizing printers, stand by
I0915 09:52:14.065334    9910 printermanager.go:283] Registered KONICA-MINOLTA-bizhub-754-MeGehee in the cloud
I0915 09:52:14.107519    9910 printermanager.go:283] Registered KONICA-MINOLTA-bizhub-C554e-MacDonald in the cloud
I0915 09:52:14.214250    9910 printermanager.go:283] Registered KONICA-MINOLTA-bizhub-C554e-Administration in the cloud
I0915 09:52:14.271931    9910 printermanager.go:283] Registered KONICA-MINOLTA-bizhub-754-Gonyea in the cloud
I0915 09:52:14.272514    9910 printermanager.go:271] Finished synchronizing 7 printers
I0915 09:52:14.282690    9910 printermanager.go:115] Registered McGehee locally
I0915 09:52:14.311690    9910 printermanager.go:115] Registered MacDonald locally
I0915 09:52:14.321887    9910 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-754-MeGehee locally
I0915 09:52:14.332088    9910 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-C554e-MacDonald locally
I0915 09:52:14.342275    9910 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-C554e-Administration locally
I0915 09:52:14.352387    9910 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-754-Gonyea locally
I0915 09:52:14.362534    9910 printermanager.go:115] Registered Computer locally
E0915 09:52:14.363737    9910 connector.go:119] Ready to rock as proxy 'OHS-CloudPrint'
Ready to rock as proxy 'OHS-CloudPrint'
I0915 09:53:14.363703    9910 printermanager.go:219] Synchronizing printers, stand by
E0915 09:53:14.383172    9910 printermanager.go:207] Sync failed while calling GetPrinters(): Failed to call cupsDoRequest() [IPP_OP_CUPS_GET_PRINTERS]: Failed to connect to CUPS server /var/run/cups/cups.sock:631 because 4097 Error in the pull function.
I0915 09:54:14.384225    9910 printermanager.go:219] Synchronizing printers, stand by
I0915 09:54:21.199464    9910 printermanager.go:252] Printers are already in sync; there are 7
I0915 09:55:21.200271    9910 printermanager.go:219] Synchronizing printers, stand by
E0915 09:55:21.219388    9910 printermanager.go:207] Sync failed while calling GetPrinters(): Failed to call cupsDoRequest() [IPP_OP_CUPS_GET_PRINTERS]: Failed to connect to CUPS server /var/run/cups/cups.sock:631 because 4097 Error in the pull function.
I0915 09:56:21.219725    9910 printermanager.go:219] Synchronizing printers, stand by
I0915 09:56:28.019341    9910 printermanager.go:252] Printers are already in sync; there are 7
I0915 09:57:28.020064    9910 printermanager.go:219] Synchronizing printers, stand by
E0915 09:57:28.039347    9910 printermanager.go:207] Sync failed while calling GetPrinters(): Failed to call cupsDoRequest() [IPP_OP_CUPS_GET_PRINTERS]: Failed to connect to CUPS server /var/run/cups/cups.sock:631 because 4097 Error in the pull function.
```

I also don't see any of these printers in my Cloud Print web interface. Well, there is one called "Computer" which shows as "Paused".  But none of the others. 

I'm sorry to need so much support. But I can't find any other good documentation on this!

**Comment by timmrogers** [comment]
Ah, I see. I'll let you know.

**Comment by jacobmarble** [comment]
Another idea. I don't have a manual either. :-/

When I test on Debian, a clean install doesn't come with CA certificates, so things like `wget https://blah` fail until I install the `ca-certificates` package. It *shouldn't* make a difference here, but it's easy to try.

**Comment by timmrogers** [comment]
That certainly helps.

```

trogers@debian:~$ CUPS_ENCRYPTION=NEVER connector -logtostderr

E0915 10:01:25.330295    9949 connector.go:33] Google Cloud Print CUPS
Connector version DEV-linux

Google Cloud Print CUPS Connector version DEV-linux

I0915 10:01:27.038907    9949 core.go:86] connected to CUPS server
/var/run/cups/cups.sock:631 encrypting NEVER

I0915 10:01:27.039893    9949 connector.go:88] SNMP enabled

I0915 10:01:27.052636    9949 avahi.go:257] Avahi client running.

I0915 10:01:28.309485    9949 printermanager.go:219] Synchronizing
printers, stand by

I0915 10:01:40.636733    9949 printermanager.go:252] Printers are already
in sync; there are 7

I0915 10:01:40.646616    9949 printermanager.go:115] Registered MacDonald
locally

I0915 10:01:40.668510    9949 printermanager.go:115] Registered Computer
locally

I0915 10:01:40.678445    9949 printermanager.go:115] Registered
KONICA-MINOLTA-bizhub-754-Gonyea locally

I0915 10:01:40.688320    9949 printermanager.go:115] Registered
KONICA-MINOLTA-bizhub-C554e-MacDonald locally

I0915 10:01:40.698199    9949 printermanager.go:115] Registered
KONICA-MINOLTA-bizhub-754-MeGehee locally

I0915 10:01:40.708040    9949 printermanager.go:115] Registered
KONICA-MINOLTA-bizhub-C554e-Administration locally

I0915 10:01:40.717954    9949 printermanager.go:115] Registered McGehee
locally

E0915 10:01:40.719039    9949 connector.go:119] Ready to rock as proxy
'OHS-CloudPrint'

Ready to rock as proxy 'OHS-CloudPrint'

I0915 10:02:40.719012    9949 printermanager.go:219] Synchronizing
printers, stand by

I0915 10:02:47.139675    9949 printermanager.go:252] Printers are already
in sync; there are 7

```

I'll try installing ca-certificates.

I want to ask about sharing as well...but give me a second...

Thanks.

**Comment by timmrogers** [comment]
That certainly makes it happier.

```
trogers@debian:~$ CUPS_ENCRYPTION=NEVER connector -logtostderr
E0915 10:01:25.330295    9949 connector.go:33] Google Cloud Print CUPS Connector version DEV-linux
Google Cloud Print CUPS Connector version DEV-linux
I0915 10:01:27.038907    9949 core.go:86] connected to CUPS server /var/run/cups/cups.sock:631 encrypting NEVER
I0915 10:01:27.039893    9949 connector.go:88] SNMP enabled
I0915 10:01:27.052636    9949 avahi.go:257] Avahi client running.
I0915 10:01:28.309485    9949 printermanager.go:219] Synchronizing printers, stand by
I0915 10:01:40.636733    9949 printermanager.go:252] Printers are already in sync; there are 7
I0915 10:01:40.646616    9949 printermanager.go:115] Registered MacDonald locally
I0915 10:01:40.668510    9949 printermanager.go:115] Registered Computer locally
I0915 10:01:40.678445    9949 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-754-Gonyea locally
I0915 10:01:40.688320    9949 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-C554e-MacDonald locally
I0915 10:01:40.698199    9949 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-754-MeGehee locally
I0915 10:01:40.708040    9949 printermanager.go:115] Registered KONICA-MINOLTA-bizhub-C554e-Administration locally
I0915 10:01:40.717954    9949 printermanager.go:115] Registered McGehee locally
E0915 10:01:40.719039    9949 connector.go:119] Ready to rock as proxy 'OHS-CloudPrint'
Ready to rock as proxy 'OHS-CloudPrint'
I0915 10:02:40.719012    9949 printermanager.go:219] Synchronizing printers, stand by
I0915 10:02:47.139675    9949 printermanager.go:252] Printers are already in sync; there are 7
```

Which leads met

**Comment by timmrogers** [comment]
ca-certificates was already installed. So that isn't the issue.

OK, aside from this issue, I'm afraid that I likely don't understand something basic.

I've used Google Cloud print printers both the the native firmware (on Brother printers) and with the old Chrome hack on Linux. The firmware on the Brother seems flaky - it shows the printer as offline after a few weeks and requires that it be unregistered, re-registered, and re-shared to get it working again. 

That is why I am here! 

My goal is to share 5 or 6 printers with a groups of students. But I'm not quite sure how to do that. I'm clearly missing something.

Thanks for all the help. I do appreciate it.

**Comment by jacobmarble** [comment]
First, you can make that change permanent by modifying `/etc/cups/client.conf`. Or just add the `CUPS_ENCRYPTION` prefix to your startup script.

Second, according to the text you pasted above, you have registered 7 printers on both the local network, as well as in "the cloud".
- If your group of students print from the same subnet as the printer, then you're done. Any Google Cloud Print client (built in to Chrome, ChromeOS, Android; installable on iOS and Windows) should discover the printers as of now.
- If your group of students print from a different subnet, then go to https://www.google.com/cloudprint/#printers, select a printer, and click the green share button. The rest is intuitive.

**Comment by timmrogers** [comment]
So we seem to be getting closer...but no joy yet! :)

I can see the Brother printer here in the Computer lab. Yippee! However, it shows in the web-based Cloud Print interface as "Paused", printing to it does not work, and the connector shows:

```
trogers@debian:~$ CUPS_ENCRYPTION=NEVER connector -logtostderr
E0915 11:08:07.688726   10095 connector.go:33] Google Cloud Print CUPS Connector version DEV-linux
Google Cloud Print CUPS Connector version DEV-linux
I0915 11:08:09.459001   10095 core.go:86] connected to CUPS server /var/run/cups/cups.sock:631 encrypting NEVER
I0915 11:08:09.460020   10095 connector.go:88] SNMP enabled
I0915 11:08:09.472024   10095 avahi.go:257] Avahi client running.
I0915 11:08:10.188116   10095 printermanager.go:219] Synchronizing printers, stand by
I0915 11:08:11.149120   10095 printermanager.go:283] Registered Computer in the cloud
I0915 11:08:11.553955   10095 printermanager.go:289] Shared Computer
I0915 11:08:11.554735   10095 printermanager.go:271] Finished synchronizing 4 printers
I0915 11:08:11.565535   10095 printermanager.go:115] Registered McGehee locally
I0915 11:08:11.575735   10095 printermanager.go:115] Registered MacDonald locally
I0915 11:08:11.585867   10095 printermanager.go:115] Registered Brother-HL-5450DN-Carriage-House locally
I0915 11:08:11.595983   10095 printermanager.go:115] Registered Computer locally
E0915 11:08:11.597158   10095 connector.go:119] Ready to rock as proxy 'OHS-CloudPrint'
Ready to rock as proxy 'OHS-CloudPrint'
I0915 11:08:33.619441   10095 printermanager.go:507] Received job 0b8fe226-96c4-85b9-5e03-3fe7d5fcd11a
I0915 11:08:34.300600   10095 printermanager.go:484] Downloaded job 0b8fe226-96c4-85b9-5e03-3fe7d5fcd11a in 534.163542ms
E0915 11:08:34.304742   10095 printermanager.go:556] Failed to send job 0b8fe226-96c4-85b9-5e03-3fe7d5fcd11a to CUPS: Failed to call cupsPrintFile2(): 1286 Destination "Computer" is not accepting jobs.
```

Printing to that  printer via a normal network connection from my Mac works fine.

Thanks.

**Comment by jacobmarble** [comment]
Can you print from the debian machine with the lp command? That would
exercise CUPS, which seems to believe the printer is stopped.

Sometimes you can fix that at http://deian-server-ip-address:631 , which
might be disabled. You can enable it with `cupsctl --remote-admin`.

**Comment by timmrogers** [comment]
I suspect that the problem is at the CUPS level. I'll see if I can get that sorted before I bother you again! 

Thanks again.

**Comment by jacobmarble** [comment]
I suspect the same. Good luck!

**Comment by jacobmarble** [comment]
Did you figure the stopped/paused problem Tim?

**Comment by timmrogers** [comment]
Not quite.

There is certainly an issue with CUPS (or, more likely, with my
understanding). And I've made some progress - I did get one of the printers
working, and I did find that at least one of the issues was with A4 vs
Letter sized paper and the defaults in CUPS.

But I still need to use your workaround for the "Error in the pull
function" issue
and disable encryption. And the web interface also seems to have encryption
issues, it works for http:// but fails when it tries to move to https://

So - I am getting closer, but it isn't ready to make available to the
students. I'll keep at it!

Thanks!
